### PR TITLE
Implement VerifyImports

### LINKED ISSUE #22: Implement verifyimports

### PR DESCRIPTION
**Pull request by jakelang**
Closes #22.

### CODE REVIEW

File: libchisel/src/verifyimports.rs [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            entries: vec![
+                ImportType::Function(
+                    "ethereum",
+                    "useGas",
+                    FunctionType::new(vec![ValueType::I64], None),
+                ),
+                ImportType::Function(
+                    "ethereum",
+                    "getGasLeft",
+                    FunctionType::new(Vec::new(), Some(ValueType::I64)),
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: Cannot you use `vec![]` here?

Comment (jakelang) [reply libchisel/src/verifyimports.rs]: yeah. not much of a difference, I would say.

**Comment by axic** [comment]
Don't forget to update the readme.

File: README.md [hunk 2]
```diff
@@ -20,14 +20,19 @@ This comes with some presets:
...
+
+The following presets are provided:
+- `ewasm`: Verifies the ewasm [EEI](https://github.com/ewasm/design/blob/master/eth_interface.md). Disallows unlisted imports, and does not require that the entire interface be imported.
 
 ### verifyexports
 
-Validates that the module's exports are compliant with the given export interface.
-Can be set to allow or prohibit additional exports that are not listed.
+Verifies that the module's exports are compliant with the provided export interface.
+Can be set to allow or prohibit unlisted additional imports.
```
[the remark this answers is not in the document]
Comment (jakelang) [reply README.md]: oh i guess i accidentally thought i mixed these up in the diff. fixing now

File: libchisel/src/verifyimports.rs (hunk 1 above)
Comment (axic) [reply libchisel/src/verifyimports.rs]: Though it is more consistent with the other inline declaration of `vec`.

Comment (jakelang) [reply libchisel/src/verifyimports.rs]: true, will fix

Comment (jakelang) [reply libchisel/src/verifyimports.rs]: fixed.

File: libchisel/src/verifyimports.rs [hunk 3]
```diff
@@ -0,0 +1,883 @@
...
+pub struct VerifyImports<'a> {
+    entries: Vec<ImportType<'a>>,
+    require_all: bool,
+    allow_unlisted: bool,
+}
+
+impl<'a> VerifyImports<'a> {
+    fn ewasm() -> Self {
+        VerifyImports {
+            //FIXME: Do not inline this
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: What is this referring to?

File: libchisel/src/verifyimports.rs [hunk 4]
```diff
@@ -0,0 +1,883 @@
...
+
+/// Struct on which ModuleValidator is implemented.
+pub struct VerifyImports<'a> {
+    entries: Vec<ImportType<'a>>,
+    require_all: bool,
+    allow_unlisted: bool,
+}
+
+impl<'a> VerifyImports<'a> {
+    fn ewasm() -> Self {
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: Some other modules use `with_preset(preset: &str)`, which is the preferred one?

File: libchisel/src/verifyimports.rs [hunk 5]
```diff
@@ -0,0 +1,883 @@
...
+/// Trait over ImportType that checks an import's other properties in the case that it is imported.
+trait ImportCheck {
+    fn check(&self, module: &Module) -> ImportStatus;
+}
+
+/// Struct on which ModuleValidator is implemented.
+pub struct VerifyImports<'a> {
+    entries: Vec<ImportType<'a>>,
+    require_all: bool,
+    allow_unlisted: bool,
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: Can we document these fields here for rustdoc ?

File: libchisel/src/verifyimports.rs (hunk 4 above)
Comment (jakelang) [reply libchisel/src/verifyimports.rs]: I think that `with_preset` would be better. will fix

File: libchisel/src/verifyimports.rs (hunk 3 above)
Comment (jakelang) [reply libchisel/src/verifyimports.rs]: It is messy to inline all the EEI func sigs in the constructor.

File: libchisel/src/verifyimports.rs (hunk 5 above)
Comment (jakelang) [reply libchisel/src/verifyimports.rs]: Will do

File: libchisel/src/verifyimports.rs (hunk 4 above)
Comment (axic) [reply libchisel/src/verifyimports.rs]: Actually it turns out only one module uses that, the others do `ewasm`.

Comment (jakelang) [reply libchisel/src/verifyimports.rs]: fuck lol, i just fixed it. should i revert?

File: libchisel/src/verifyimports.rs (hunk 5 above)
Comment (jakelang) [reply libchisel/src/verifyimports.rs]: done.

File: libchisel/src/verifyimports.rs [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+    fn check(&self, module: &Module) -> ImportStatus;
+}
+
+/// Struct on which ModuleValidator is implemented.
+pub struct VerifyImports<'a> {
+    /// List of function signatures to check.
+    entries: Vec<ImportType<'a>>,
+    /// Option to require the presence of all listed imports in the module. When false, only the
+    /// validity of existing imports on the list is checked.
+    require_all: bool,
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: What use case do you imagine?

File: libchisel/src/verifyimports.rs [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+    /// Option to allow imports that are not listed in `entries`.
+    allow_unlisted: bool,
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: Also what is the usecase of this?

Comment (jakelang) [reply libchisel/src/verifyimports.rs]: when you don't want extra imports. /s

I'm going to assume you mean true, in which case something like wasmchain might have imports that correspond to exposed functions in other contracts, which may not be immediately available to chisel.

File: libchisel/src/verifyimports.rs (hunk 6 above)
Comment (jakelang) [reply libchisel/src/verifyimports.rs]: Interfaces of mandatory callbacks.
Useful for ensuring contracts import something like `return`, or a family of such functions that are necessary for the contract to be valid.

File: libchisel/src/verifyimports.rs [hunk 8]
```diff
@@ -0,0 +1,898 @@
...
+            0
+        };
+
+        Ok(match (self.require_all, self.allow_unlisted) {
+            (true, true) => !self
+                .entries
+                .iter()
+                .map(|e| e.is_imported(module))
+                .find(|e| *e == false)
+                .is_some(),
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: Instead of `is_some` and negation is there any other option?

File: libchisel/src/verifyimports.rs [hunk 9]
```diff
@@ -0,0 +1,898 @@
...
+impl<'a> ModuleValidator for VerifyImports<'a> {
+    fn validate(self, module: &Module) -> Result<bool, String> {
+        let import_section_len = if let Some(section) = module.import_section() {
+            section.entries().len()
+        } else {
+            0
+        };
+
+        Ok(match (self.require_all, self.allow_unlisted) {
+            (true, true) => !self
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: For consistency can you also wrap this into curly brackets?

File: libchisel/src/verifyimports.rs [hunk 10]
```diff
@@ -0,0 +1,898 @@
...
+                self
+                    .entries
+                    .iter()
+                    .map(|e| e.is_imported(module))
+                    .find(|e| *e == false)
+                    .is_none()
+                    && (self.entries.len() == import_section_len)
+            }
+            (false, true) => self
+                .entries
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: This case basically doesn't care what the imports are, except that the listed ones must match.

File: libchisel/src/verifyimports.rs [hunk 11]
```diff
@@ -0,0 +1,898 @@
...
+                // equal to the number of valid imports, all existent imports are valid; If all
+                // existent imports are valid, there are no existent invalid imports; qed
+                valid_entries_count == import_section_len
+            }
+        })
+    }
+}
+
+impl<'a> IsImported for ImportType<'a> {
+    fn is_imported(&self, module: &Module) -> bool {
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: Need to document that this also checks the signature.

File: libchisel/src/verifyimports.rs [hunk 12]
```diff
@@ -0,0 +1,898 @@
...
+            match self {
+                ImportType::Function(namespace, field, sig) => {
+                    has_func_import(module, namespace, field, sig)
+                }
+                ImportType::Global(namespace, field) => {
+                    has_global_import(section, namespace, field)
+                }
+                ImportType::Memory(namespace, field) => {
+                    has_memory_import(section, namespace, field)
+                }
```
Comment (axic) [inline libchisel/src/verifyimports.rs]: I guess these don't necessarily need the curly braces above.

**Pull request merged**